계산기 기능개선

jjori-master opened the pull request:
- interface map을 활용한 if구문 제거
- main method추가Scanner클래스를 활용한 콘솔입력

jjori-master committed:
계산기 기능개선
- interface map을 활용한 if구문 제거
- main method추가Scanner클래스를 활용한 콘솔입력

javajigi commented on src/main/java/calculator/Calculator.java:
```diff
@@ -1,76 +1,87 @@
-            }
+            result = operate(operator, result, Integer.parseInt(param));
+            operator = null;
+        }
```
메소드가 너무 많은 일을 하고 있다. 다음 원칙을 지켜 메소드를 구현한다.

작게 만들어라. - 함수를 만드는 첫 번재 규칙은 '작게'다. 함수를 만드는 두 번째 규칙은 '더 작게'다.
한 가지만 해라. - 함수는 한 가지를 해야 한다. 그 한 가지를 잘 해야 한다. 그 한 가지만 해야 한다.

javajigi commented on src/main/java/calculator/Calculator.java:
```diff
@@ -1,76 +1,87 @@
+        map.put("/", new Division());
 
-        return result;
+        return map.get(operator).operate(a, b);
```
구현 잘 했네요. 💯

javajigi reviewed (changes requested):
이 pr에 RacingCar 1단계 추가로 구현해서 add, commit, push하고 다시 리뷰 요청해 주세요.

jjori-master committed:
자동차 경주 게임 1단계

javajigi commented on src/main/java/racing/RacingGame.java:
```diff
@@ -0,0 +1,72 @@
+
+    public RacingGame(Speed speed) {
+        this.speed = speed;
+        time = 5;
```
값을 하드코딩하지 마라.
상수(static final)를 만들고 이름을 부여해 이 변수의 역할이 무엇인지 의도를 드러내라.

javajigi commented on src/main/java/racing/RacingGame.java:
```diff
@@ -0,0 +1,72 @@
+    public RacingGame(Speed speed) {
+        this.speed = speed;
+        time = 5;
+        setCarPositions(3);
```
this(speed, 5, 3) 과 같이 구현하면 생성자의 중복을 제거할 수 있다.

javajigi commented on src/main/java/racing/RacingGame.java:
```diff
@@ -0,0 +1,72 @@
+        setCarPositions(numberOfCar);
+    }
+
+    public void setCarPositions(int numberOfCar) {
```
이 메소드가 public인 이유는? private 이어도 되지 않을까?

jjori-master replied:
@javajigi 메소드가 private일 경우 단일 테스트가 불가능 한데요
생성자 호출시 setCarPositions 함수를 바로 호출하니 생성자 테스트와 같이 묶어서 테스트를 해야 하는건지 알고 싶습니다.

javajigi replied:
@jjori-master 테스트를 위해 오픈해야 된다면 public보다는 default(접근 제어자가 없는 메소드)로 구현할 것을 추천해요.

javajigi commented on src/main/java/racing/RacingGame.java:
```diff
@@ -0,0 +1,72 @@
+        System.out.println(snapshot);
+    }
+
+    public String getDistance(int position) {
```
인스턴스 변수에 접근하고 있지 않은 클래스(static) 메소드로 구현해도 되지 않을까?

jjori-master replied:
@javajigi 제가 아직 static으로 선언하면 좋은(?) 함수와 그렇지 않은 함수를 잘 선별하지 못해서 인식하지 못하고 static으로
선언하지 못했습니다. static 선언과 관련하여 검색하던 중
["왜 자바에서 static의 사용을 지양해야 하는가?"](http://tech.thegajago.com/2016/02/20/%EC%99%9C-%EC%9E%90%EB%B0%94%EC%97%90%EC%84%9C-static%EC%9D%98-%EC%82%AC%EC%9A%A9%EC%9D%84-%EC%A7%80%EC%96%91%ED%95%B4%EC%95%BC-%ED%95%98%EB%8A%94%EA%B0%80/l) 글을 살펴보게 되었습니다.

static을 사용하면 메모리 관리의 장점을 얻을 수 있다는 교과서적인 내용과 달리 메모리 사용에 있어서, 더 불리하고 객체 지향적이지 않다, 재사용성이 떨어진다. 라는 토론내용이 있는데, 교수님 생각은 어떠신지 알고 싶습니다.

javajigi replied:
@jjori-master 글이 사라졌는데요. 이 이야기는 인스턴스를 생성할 때 singleton 패턴의 사용을 자제하라는 것입니다. static 사용을 무조건 하지 않는 것이 맞는 것은 아니고요. 이 메소드와 같이 인스턴스 변수에 접근하지 않는 메소드는 static으로 구현하는 것이 더 낫다고 생각해요.

javajigi commented on src/main/java/racing/RacingGame.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    public boolean canIGo() {
+        return speed.fullAccel() >= 4;
```
값을 하드코딩하지 마라.
상수(static final)를 만들고 이름을 부여해 이 변수의 역할이 무엇인지 의도를 드러내라.

javajigi commented on src/test/java/racing/TestRacingGame.java:
```diff
@@ -0,0 +1,120 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class TestRacingGame {
```
RacingGameTest로 구현하는 것이 관례임

javajigi commented on src/test/java/racing/TestRacingGame.java:
```diff
@@ -0,0 +1,120 @@
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+                return ((count++) % 2) == 0 ? 3 : 4;
+            }
+        });
```
mockito를 쓰지 않고 테스트 가능하도록 구현한다.
mockito를 쓰면 테스트가 복잡해짐. 현재 구조로도 mockito 없이 테스트 가능함.

익명 클래스 활용해 처리한다.

javajigi commented on src/test/java/racing/TestRacingGame.java:
```diff
@@ -0,0 +1,120 @@
+        List<Integer> carPositions = new ArrayList();
+        carPositions.add(new Integer(1));
+        carPositions.add(new Integer(1));
+        carPositions.add(new Integer(1));
```
List<Integer> carPositions = Arrays.asList(1, 1, 1);

위와 같이 테스트 데이터 생성.

javajigi reviewed (changes requested):
인터페이스를 추출한 것은 좋은데요. 테스트 활용시 좀 더 적극적으로 활용하면 좋겠어요.

jjori-master committed:
자동차 경주 게임 1단계 1차리뷰 후처리
- 계산기 calc 메소드 로직 간소화
- 하드코딩 상수화
- 생성자 중복 제거
- mock 익명클래스로 대체 테스트
- 인스턴스 함수 일부 전역함수로 변경

jjori-master committed:
테스트 케이스 추가

javajigi reviewed (changes requested):
도메인 로직과 View 로직을 구분하기 위해 다음과 같은 구조로 도전해 보면 어떨까요?

```
public static void main(String[] args) {
    int carNo = InputView.getCarNo();
    int tryNo = InputView.getTryNo();

    RacingGame racingGame = new RacingGame(carNo);
    for(int i = 0; i < tryNo; i++) {
        GameResult result = racingGame.move();
        ResultView.print(result);
    }
}
```

GameResult에 이동한 위치 정보가 포함되어 있고요. 이 위치 정보를 활용해 출력하는 구조로 구현해 보면 좋을 것 같아요.

위 코드에서 힌트를 얻어 본인의 코드에 개선할 부분이 있는지 검토해 보고 리팩토링 해보세요.

Commit:
직장에서 한 작업 커밋

jjori-master committed:
자동차 경주 게임 1단계 리뷰후 코드 수정
- InputView, ResultView 클래스를 통해 입력과 출력기능을 분리
- MoveResult 클래스를 통해 이동후 결과 정보를 관리
- Random 테스트를 고정 데이터로 선 테스트 후 주석처리
